### ISSUE TITLE
imbalanced data

### ISSUE THREAD

**Issue by kike-0304** [basf/mamba-tabular#385]
Hi,

I am using DeepTab for a binary classification task with highly imbalanced data.

For example:

* Class 0: 90%
* Class 1: 10%

How can I apply class weighting during training?

In LightGBM, I would typically use `scale_pos_weight`, and in PyTorch I can use `BCEWithLogitsLoss(pos_weight=...)`.

Does DeepTab provide a built-in way to handle class imbalance, such as:

* class weights
* sample weights
* weighted loss functions
* weighted sampling

If so, could you provide an example?

Thank you.

**Comment by mkumar73** [comment]
Hello @kike-0304,

Thank you for the question.

To answer directly: class weights, sample weights, weighted loss functions, and weighted sampling are not currently exposed through the public estimator API in the current v1.x release.

## Current Workaround

For now, the recommended workaround is to handle imbalance before fitting the model. For numerical features, you can use resampling, for example:

```python
from imblearn.over_sampling import SMOTE
from deeptab.models import MambularClassifier

X_res, y_res = SMOTE(random_state=42).fit_resample(X_train, y_train)

model = MambularClassifier()
model.fit(X_res, y_res, max_epochs=100)
```

## What's Coming in v2.0.0

v2.0.0 will add first-class support for several of these. Based on the direction of the project:

**Weighted loss** (`pos_weight`, equivalent to `BCEWithLogitsLoss(pos_weight=...)`):

```python
import torch
import torch.nn as nn
from deeptab.models import MambularClassifier

pos_weight = torch.tensor([(y_train == 0).sum() / (y_train == 1).sum()])

# Expected v2.0.0 API (not yet available)
model = MambularClassifier(
    loss_fct=nn.BCEWithLogitsLoss(pos_weight=pos_weight)
)
model.fit(X_train, y_train, max_epochs=100)
```

**Configurable stratification** via `SplitConfig`:

```python
# Expected v2.0.0 API (not yet available)
from deeptab.configs import SplitConfig

model = MambularClassifier(
    split_config=SplitConfig(stratify=True, val_size=0.15, random_state=42)
)
```

Class weights, sample weights, and weighted sampling aren't in v2.0.0 yet. We will look into whether they can be included before the release, which is planned for later this month. Thank you!

**Comment by mkumar73** [comment]
Hi @kike-0304 

Yes, DeepTab now handles class imbalance directly in `fit()`, so you don't need to resample by hand. This is available in DeepTab v2.0.0, so make sure you are on that version. There are four options and you can combine a loss based one with a sampling based one.

The quickest thing to try is scikit-learn style class weighting:

```python
from deeptab.models import MambularClassifier

model = MambularClassifier()
model.fit(X_train, y_train, class_weight="balanced", max_epochs=100)
```

For a binary target this becomes a `pos_weight` on `BCEWithLogitsLoss` (same idea as `scale_pos_weight` in LightGBM), and for multiclass it becomes the `weight` on `CrossEntropyLoss`. You can also pass explicit weights as a dict `{0: 1.0, 1: 5.0}` or an array ordered like `np.unique(y)`.

If you want your exact PyTorch setup, pass the loss straight through and it takes priority over `class_weight`:

```python
import torch
import torch.nn as nn

pos_weight = torch.tensor([(y_train == 0).sum() / (y_train == 1).sum()])
model.fit(X_train, y_train, loss_fct=nn.BCEWithLogitsLoss(pos_weight=pos_weight))
```

Under heavy imbalance like your 90/10 case, focal loss is usually the single best setting, and you can layer class weights on top:

```python
model.fit(X_train, y_train, loss_fct="focal", class_weight="balanced")
```

If you'd rather rebalance the batches instead of the loss, there's weighted sampling too:

```python
# oversample the minority class into class balanced batches
model.fit(X_train, y_train, balanced_sampler=True)

```

So to map it to your question: class weights, sample weights, weighted loss, and weighted sampling are all covered. `class_weight="balanced"` is a good first try, and `loss_fct="focal"` tends to help most on strongly skewed data.

Hope it addresses your question and concern :)
